api/tests/Feature/RolePermissionTest.php

```diff
@@ -0,0 +1,195 @@
+
+        $this->assertTrue($this->user->hasRole('guest'));
+        $this->assertTrue($this->user->isAbleTo([
+            'view-any-classification',
+            'view-any-department',
+            'view-any-skill',
+            'view-any-skillFamily',
+            'view-any-publishedPoolAdvertisement',
+            'view-any-applicantCount',
+            'create-any-searchRequest',
+            'view-any-team'
+        ], true));
+
+        $this->cleanup();
+    }
+
+    /**
+     * Test the Base User Role
+     *
+     * @return void
+     */
+    public function test_base_user_role()
+    {
+        $baseUserRole = Role::where('name', 'base_user')->sole();
+        $this->user->attachRole($baseUserRole);
+
+        $this->assertTrue($this->user->hasRole('base_user'));
+        $this->assertTrue($this->user->isAbleTo([
+            'view-own-user',
+            'update-own-user',
+        ], true));
+
+        $this->cleanup();
+    }
+
+    /**
+     * Test the Applicant Role
+     *
+     * @return void
+     */
+    public function test_applicant_role()
+    {
+        $applicantRole = Role::where('name', 'applicant')->sole();
+        $this->user->attachRole($applicantRole);
+
+        $this->assertTrue($this->user->hasRole('applicant'));
+        $this->assertTrue($this->user->isAbleTo([
+            'view-own-application',
+            'submit-own-application',
+            'create-own-draftApplication',
+            'delete-own-draftApplication',
+            'archive-own-submittedApplication',
+        ], true));
+
+        $this->cleanup();
+    }
+
+    /**
+     * Test the Team Admin Role
+     *
+     * @return void
+     */
+    public function test_team_admin_role()
+    {
+        $teamAdminRole = Role::where('name', 'team_admin')->sole();
+        $this->user->attachRole(
+            $teamAdminRole,
+            $this->ownedTeam
+        );
+
+        $permissionsToCheck = [
+            'view-any-userBasicInfo',
+            'view-team-pool',
+            'create-team-pool',
+            'publish-team-pool',
+            'update-team-draftPool',
+            'delete-team-draftPool',
+            'update-team-poolClosingDate',
+            'view-team-submittedApplication',
+            'view-team-applicantProfile',
+            'update-team-applicationStatus',
+            'view-team-searchRequest',
+            'update-team-searchRequest',
+            'delete-team-searchRequest',
+            'view-team-teamMembers',
+            'update-team-team',
+            'view-any-role',
+            'assign-team-role',
+        ];
+
+        $this->assertTrue($this->user->hasRole('team_admin',  $this->ownedTeam));
+        $this->assertTrue($this->user->isAbleTo($permissionsToCheck, $this->ownedTeam, true));
+
+        $this->assertFalse($this->user->hasRole('team_admin',  $this->unownedTeam));
+        $this->assertFalse($this->user->isAbleTo($permissionsToCheck, $this->unownedTeam));
```

Comment:
So basically what I'm worried about, in regards to the team_check_strict, is that we won't have a team to pass in when we check `view-any-role`. Like this:
```suggestion
        $teamPermissionsToCheck = [
            'view-any-userBasicInfo',
            'view-team-pool',
            'create-team-pool',
            'publish-team-pool',
            'update-team-draftPool',
            'delete-team-draftPool',
            'update-team-poolClosingDate',
            'view-team-submittedApplication',
            'view-team-applicantProfile',
            'update-team-applicationStatus',
            'view-team-searchRequest',
            'update-team-searchRequest',
            'delete-team-searchRequest',
            'view-team-teamMembers',
            'update-team-team',
            'assign-team-role',
        ];
        $otherPermissionsToCheck = [
            'view-any-role',
        ];

        $this->assertTrue($this->user->isAbleTo($otherPermissionsToCheck, null, true));

        $this->assertTrue($this->user->hasRole('team_admin',  $this->ownedTeam));
        $this->assertTrue($this->user->isAbleTo($teamPermissionsToCheck, $this->ownedTeam, true));

        $this->assertFalse($this->user->hasRole('team_admin',  $this->unownedTeam));
        $this->assertFalse($this->user->isAbleTo($teamPermissionsToCheck, $this->unownedTeam));
```
Here's the trouble... I thought this would fail with teams_strict_check=true, and pass with teams_strict_check=false. But I've tested this locally, changing the config file, and it seems to fail either way! 😧

Reply:
haha yeah, that is part of my confusion! It seems to return the same value regardless of what you put in the strict check arg 😖

Reply:
Essentially when it is:

-  `true` you _must_ pass in the team to the role/permission check if it comes from a team other wise, it will check permissions that are _not_ associated with that team
- `false` means it does not care if the permission is attached to a team or not, it will return true no matter if that role/permission is part of a team or not and doesn't even care which team that role/permission comes from

Reply:
> It seems to return the same value regardless of what you put in the strict check arg 

I think you have to change teams_strict_check in the config file, I don't think there's an arg for it. The third arg to isAbleTo is about whether to do an ALL or ANY operation on arrays of permissions.

Reply:
> > It seems to return the same value regardless of what you put in the strict check arg
> 
> I think you have to change teams_strict_check in the config file, I don't think there's an arg for it. The third arg to isAbleTo is about whether to do an ALL or ANY operation on arrays of permissions.

Oh yeah, poor word choice on my part. I was updating config. I was running into some permission issues on WSL where the config cache was not updating. I'm pretty sure I have it sorted out now.

Do we want it to be `false` still? This might cause problems when we start to mix team and non-team roles. We will need to be careful to _always_ pass a team in when checking elevated team permissions since a check with a `null` team checks all teams for that permission rather than all permissions that are not associated with a team.